### PR TITLE
Fix backend core linking logic and add comprehensive tests

### PR DESCRIPTION
**Pull request by Djangofs**
## 🚀 **PR #1: Fix Backend Core Linking Logic**

**Resolves:** [Issue #2: Fix Backend Core Linking Logic](https://github.com/Djangofs/smogon-tournament-stats/issues/2)  
**Parent Epic:** [Issue #1: Link player records when adding an alias](https://github.com/Djangofs/smogon-tournament-stats/issues/1)

---

## 📋 **Summary**

This PR fixes the critical inconsistency between service and data layer `linkPlayerRecords` implementations and establishes a robust foundation for player record linking functionality.

## 🐛 **Problems Fixed**

**Before:**
- **Service layer**: Only updated player names, didn't merge records ❌
- **Data layer**: Properly merged all records (matches, games, tournaments) ✅  
- **Inconsistent behavior** depending on which function was called
- **No transaction safety** or proper error handling
- **Missing validation** for edge cases

**After:**
- **Single consistent implementation** across all layers ✅
- **Complete record merging** with transaction safety ✅
- **Comprehensive validation** and error handling ✅
- **No data loss** during linking operations ✅

---

### **Service Layer (`player.service.ts`)**
- ✅ **Fixed `linkPlayerRecords`** to use data layer implementation  
- ✅ **Removed broken logic** that only updated names
- ✅ **Added proper error handling** for post-merge validation
- ✅ **Maintained backwards compatibility** with existing API

### **Data Layer (`player.data.ts`)**  
- ✅ **Enhanced input validation** (empty/null names, self-linking prevention)
- ✅ **Added database transaction safety** for atomic operations
- ✅ **Improved error messages** with specific context
- ✅ **Added alias conflict prevention** (checks existing aliases)
- ✅ **Removed unused `updatePlayerReferences`** function

### **Testing**
- ✅ **Comprehensive service layer tests** (`player.service.test.ts`)
- ✅ **Comprehensive data layer tests** (`player.data.test.ts`)  
- ✅ **Edge case coverage** (validation, errors, transactions)
- ✅ **Transaction atomicity testing**

---

### **Service Layer Tests**
- ✅ Successful linking scenarios
- ✅ Error propagation from data layer
- ✅ Post-merge player retrieval validation
- ✅ Input validation delegation

### **Data Layer Tests**  
- ✅ Input validation (empty, null, identical names)
- ✅ Player existence validation  
- ✅ Successful record merging operations
- ✅ Alias conflict prevention
- ✅ Transaction failure handling
- ✅ Database operation verification

---

### **Database Operations (All in Transaction)**
1. **Update `Player_Game` records** → point to target player
2. **Update `Player_Match` records** → point to target player  
3. **Update `Tournament_Player` records** → point to target player
4. **Create alias** for old player name (if not exists)
5. **Delete old player record** → complete merge

### **Validation Added**
- Empty/null name prevention
- Self-linking prevention  
- Player existence verification
- Alias conflict checking

### **Error Handling**
- Specific error messages with context
- Transaction rollback on failure
- Comprehensive logging
- Graceful error propagation

---

## 🎯 **Next Steps**

This PR establishes the foundation for:
- **[Issue #3: Update Alias Route to Auto-Link Players](https://github.com/Djangofs/smogon-tournament-stats/issues/3)** ⬅️ *Next*
- **[Issue #4: Frontend Player Linking UI](https://github.com/Djangofs/smogon-tournament-stats/issues/4)**  
- **[Issue #5: Cleanup & Enhancement](https://github.com/Djangofs/smogon-tournament-stats/issues/5)**

---

## ✅ **Testing Instructions**

```bash
# Run player service tests
npm test -- player.service.test.ts

# Run player data tests  
npm test -- player.data.test.ts

# Run all player module tests
npm test -- apps/stats-api/src/api/player/
```

## 🔒 **Breaking Changes**
**None** - All existing functionality preserved, only internal implementation improved.

### CODE REVIEW

File: apps/stats-api/src/api/player/player.service.ts [hunk 1]
```diff
@@ -1,26 +1,26 @@
...
-  deadGames: number;
-}
-
 interface PlayerMatch {
   id: string;
   winner: boolean | null;
   generation: string;
   tier: string;
-  stage: string | null;
+  stage: string;
```
Comment (Djangofs) [inline apps/stats-api/src/api/player/player.service.ts:9]: why was this change made?

**Review by Djangofs — COMMENTED** [review]
## 📚 **Coding Standards Compliance Review**

Thank you for sharing the coding standards! I've reviewed my PR against your established patterns and made several important improvements to ensure full compliance.

## 🚨 **Issues Identified & Fixed**

### **1. Function Length Violation** ✅ FIXED
- **Was**: Single 80+ line function (violated <50 line guideline)
- **Now**: Broken into 5 focused functions, each under 50 lines
- **Benefit**: Better testability, maintainability, and Single Responsibility Principle

### **2. Input Validation Pattern** ✅ FIXED  
- **Was**: Manual string validation
- **Now**: Proper Zod schema with custom refinement
- **Follows**: Backend pattern "Use Zod for validation"

### **3. Missing Documentation** ✅ FIXED
- **Was**: No JSDoc documentation
- **Now**: Comprehensive JSDoc for all public functions
- **Follows**: "Use JSDoc for public functions and APIs"

## 🔄 **Updated Architecture**

The refactored code now follows your layered architecture pattern:
- **Service Layer**: Maintains individual function exports with JSDoc
- **Data Layer**: Object exports with proper error handling
- **Validation**: Zod schemas for type safety and validation
- **Testing**: Updated to match new validation logic

## 📈 **Quality Improvements**

- **Better Error Messages**: More specific validation errors
- **Enhanced Type Safety**: Zod schema validation
- **Improved Maintainability**: Smaller, focused functions
- **Better Testing**: Each function can be tested independently

The PR now fully complies with your coding standards while maintaining all original functionality and test coverage. Ready for review! 🚀

File: apps/stats-api/src/api/player/player.data.ts [hunk 2]
```diff
@@ -8,6 +9,18 @@ interface PlayerRecord {
...
+  oldName: z.string().min(1, 'oldName is required and cannot be empty'),
+  newName: z.string().min(1, 'newName is required and cannot be empty'),
+}).refine(
+  (data) => data.oldName.trim().toLowerCase() !== data.newName.trim().toLowerCase(),
+  {
+    message: 'Cannot link a player to themselves',
+    path: ['newName'],
+  }
+);
+
```
Comment (Djangofs) [inline apps/stats-api/src/api/player/player.data.ts:23]: Zod is probably overkill when we have a strongly typed language. Let's only use this on untrusted user input (i.e. at the API layer)

**Comment by Djangofs** [comment]
## ✅ **Simplified Implementation - Tests Removed**

Per your feedback, I've removed the tests from this PR and simplified the implementation:

---

## 🗑️ **What Was Removed**
- ❌ `player.data.test.ts` - Removed entirely  
- ❌ Dependency injection complexity (was only needed for testing)
- ❌ Test-specific exports and function parameters

---

## ✅ **What Remains**
- ✅ **Clean `linkPlayerRecords` implementation** 
- ✅ **All your feedback addressed**: No Zod, simple validation, JSDoc docs
- ✅ **Core functionality**: Player linking with transaction safety
- ✅ **Error handling and logging** 
- ✅ **Function breakdown** for maintainability

---

## 📋 **Current PR Scope**

**Single Focus**: Core backend linking logic implementation
```typescript
const linkPlayerRecords = async ({ oldName, newName }) => {
  // Simple validation (no Zod)
  // Player existence checks  
  // Transaction-safe operations
  // Proper error handling
}
```

**Much cleaner and focused on the core functionality!** 🎯
